### PR TITLE
Cheeky myFT UI js refactor

### PR DESCRIPTION
**Pull request by laurieboyes**
Lots of fun was had here. Not aiming for perfection so much as I am for 'better than it was before', so any comments around the problems/decisions inherited from the old version will probably not be actioned in this PR, unless they're quick fixes. Still interested to hear general thoughts on these however. Also, I'll be extra grateful for comments on file/variable naming, which, true to reputation, was hard.

Key things:
- No longer support multiselects (now all gone).
- Split lists stuff into its own file, where it can be left to rot.
- Hide the complexity of the myFT client load event handling stuff into its own little tested lib.
- Hide the complexity of updating the myFT buttons into its own little tested lib.
- Wrote some tests (but not for everything).

The `ui/index.js` file this was all split out from was so big that its diff doesn't show up by default 😂

I was very pleased to find Inject Loader (https://github.com/plasticine/inject-loader), which allows us to mock webpacked dependencies in Karma tests :+1::+1::+1:

Still to do:

- [ ] Have a good tidy up of the `getInteractionHandler` function in `myft-buttons.js`. This is what I'll need to change to actually do the work I'm supposed to be doing for CAPI 😁
- [ ] Write more tests
- [ ] (later on maybe) Rationalise with the ui-instant code? It's probably fine left alone though

### CODE REVIEW

File: myft/test/ui/lib/button-states.spec.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+			expect(pressedValues)
+				.to.deep.equal([
+					'false',
+					'true',
+					'false'
+				])
+		});
+
+
+		it('not update the states of buttons for forms that are not of the requested relationshipName, regardless of the ID', () => {
```
Comment (keirog) [inline myft/test/ui/lib/button-states.spec.js]: +should (not update)

File: myft/test/ui/lib/loaded-relationships.spec.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	})
+
+	describe('waitForRelationshipsToLoad', () => {
+		it('should return a promise that resolves instantly if the requested relationships have already been loaded', () => {
+			mocks.myFtClient.loaded = {
+				'followed.concept': {}
+			};
+			return Promise.race([
+				loadedRelationships.waitForRelationshipsToLoad('followed'),
+				new Promise((_, reject) => setTimeout(() => reject(new Error('Not instant')), 100))
```
Comment (keirog) [inline myft/test/ui/lib/loaded-relationships.spec.js]: thisishardtoreadyo

File: myft/test/ui/personalise-links.spec.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
+/* global expect */
+
+function noWhitespace (str) {
```
Comment (keirog) [inline myft/test/ui/personalise-links.spec.js]: Nice

File: myft/test/ui/update-ui.spec.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+describe('Update UI', () => {
+
+	let updateUi;
+	let stubs;
+	let mockRelationships;
+
+	const mockUiSelectorsMap = new Map([
+		['saved', '[data-myft-ui="saved"]'],
+		['followed', '[data-myft-ui="follow"]'],
+		['danced', '[data-myft-ui="danced"]'],
```
Comment (keirog) [inline myft/test/ui/update-ui.spec.js]: 😂

File: myft/test/ui/update-ui.spec.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		})
+	})
+
+	it('should call personaliseLinks if `ignoreLinks` is not truthy', () => {
+		updateUi(document.body);
+		expect(stubs.personaliseLinksStub).to.have.been.called;
+	});
+
+	it('should not call personaliseLinks if `ignoreLinks` is truthy', () => {
+		updateUi(document.body, true);
```
Comment (keirog) [inline myft/test/ui/update-ui.spec.js]: One day we will be rid of this 🌹

File: myft/ui/myft-buttons.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+		} else if (input.type === 'hidden') {
+			meta[input.name] = input.value;
+		}
+	});
+
+	return meta;
+}
+
+function getInteractionHandler (relationshipName) {
```
Comment (keirog) [inline myft/ui/myft-buttons.js]: 😩  One day

Comment (laurieboyes) [inline myft/ui/myft-buttons.js]: One day soon!

File: myft/ui/personalise-links.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+const myFtClient = require('next-myft-client');
+const $$ = require('n-ui-foundations').$$
```
Comment (keirog) [inline myft/ui/personalise-links.js]: ;

File: myft/ui/update-ui.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+export default function (contextEl, ignoreLinks) {
+	if (!ignoreLinks) {
+		personaliseLinks(contextEl);
+	}
+
+	for (let relationshipName of uiSelectorsMap.keys()) {
+		const relationships = loadedRelationships.getRelationships(relationshipName);
+		if (relationships.length > 0) {
+			const subjectIds = relationships.map(item => item.uuid)
```
Comment (keirog) [inline myft/ui/update-ui.js]: ;

File: myft/test/ui/personalise-links.spec.js (hunk 3 above)
Comment (leggsimon) [inline myft/test/ui/personalise-links.spec.js]: could use https://github.com/i-like-robots/chai-html to assert against html strings

File: myft/ui/personalise-links.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+const myFtClient = require('next-myft-client');
+const $$ = require('n-ui-foundations').$$;
+
+export default function (el) {
```
Comment (leggsimon) [inline myft/ui/personalise-links.js]: not entirely sure what the best thing to do is but we have a mixture of `const...require`s and `export`s. Would we prefer to go all in with `import/export` or `require/module.exports`.

Not sure what the pros/cons of either are though.

File: myft/test/ui/update-ui.spec.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+		for (let relationshipName of keys) {
+			const relationships = stubs.getRelationshipsStub(relationshipName);
+			if (relationships.length > 0) {
+				const subjectIds = relationships.map(item => item.uuid);
+				stubs.setStateOfManyButtonsStub(relationshipName, subjectIds, true, contextEl);
+			}
+		}
+	}
+
+	it.only('should call `setStateOfManyButtons` with an array of subjectIds for each relationship', () => {
```
Comment (leggsimon) [inline myft/test/ui/update-ui.spec.js]: remove `.only`

File: myft/test/ui/update-ui.spec.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+		expect(stubs.personaliseLinksStub).to.have.been.called;
+	});
+
+	it('should not call personaliseLinks if `ignoreLinks` is truthy', () => {
+		updateUi(document.body, true);
+		expect(stubs.personaliseLinksStub).to.not.have.been.called;
+	});
+
+
+	function ohDear (contextEl, ignoreLinks) {
```
Comment (leggsimon) [inline myft/test/ui/update-ui.spec.js]: What is this for?!?!

File: myft/test/ui/update-ui.spec.js (hunk 5 above)
Comment (leggsimon) [inline myft/test/ui/update-ui.spec.js]: What are some cases where we do actually want to ignore the links?

File: myft/ui/update-ui.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+	if (!ignoreLinks) {
+		personaliseLinks(contextEl);
+	}
+
+	for (let relationshipName of uiSelectorsMap.keys()) {
+		const relationships = loadedRelationships.getRelationships(relationshipName);
+		if (relationships.length > 0) {
+			const subjectIds = relationships.map(item => item.uuid);
+			buttonStates.setStateOfManyButtons(relationshipName, subjectIds, true, contextEl);
+		}
```
Comment (leggsimon) [inline myft/ui/update-ui.js]: Admittedly, I'm not massively familiar with this code base but something I've observed in this file is that I have literally no idea what this is doing. I'm not sure whether this is something that could be improved by just clearer naming of things?

File: myft/ui/myft-buttons.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+		const isPressed = button.getAttribute('aria-pressed') === 'true';
+		const action = isPressed ? 'remove' : 'add';
+		const id = formEl.getAttribute(idPropertiesMap.get(relationshipName));
+		const type = typesMap.get(relationshipName);
+
+		const hiddenFields = $$('input[type="hidden"]', formEl);
+		let meta = extractMetaData([button, ...hiddenFields]);
+
+		if (~['add', 'remove'].indexOf(action)) {
```
Comment (leggsimon) [inline myft/ui/myft-buttons.js]: wat

File: karma.conf.js [hunk 14]
```diff
@@ -50,18 +56,18 @@ module.exports = function (karma) {
 						query: {
 							cacheDirectory: true,
 							presets: ['es2015', 'react'],
```
Comment (leggsimon) [inline karma.conf.js]: ⚛️ ?? 😱 🙀

File: myft/test/ui/update-ui.spec.js (hunk 5 above)
Comment (laurieboyes) [inline myft/test/ui/update-ui.spec.js]: Oh, we actually don't anywhere. Sick 👍

File: myft/test/ui/personalise-links.spec.js (hunk 3 above)
Comment (laurieboyes) [inline myft/test/ui/personalise-links.spec.js]: I really like this idea but I believe it needs a karma adapter to be able to use with karma. Sticking with this for now.

File: myft/ui/myft-buttons.js (hunk 13 above)
Comment (laurieboyes) [inline myft/ui/myft-buttons.js]: 😄  Cool right. This function is next on the list

File: myft/ui/personalise-links.js (hunk 9 above)
Comment (laurieboyes) [inline myft/ui/personalise-links.js]: 👍  Going full hipster (es6 throughout)

**Pull request merged**